### PR TITLE
Complete EventListener

### PR DESCRIPTION
**Pull request by yschimke**
Call all methods in EventListener.
Adds proxy to connectEnd to allow matching to start event unambiguously.
Test EventListener fails fast for unmatched end event.
Make RecordingEventListener a top level class.

n.b. Ordering of ResponseBodyEvent needs to be addressed in another PR.  Currently it can be delayed by client not consuming output, even if all bytes have been read off the wire.

### CODE REVIEW

**Comment by yschimke** [comment]
I'll try to finish off the TODOs this weekend, I'd like to get it in before the next release so it isn't a partial feature.  If I'm blocking anyone during the work week, please steal off me.

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java [hunk 1]
```diff
@@ -611,88 +613,175 @@ private void enableTlsWithTunnel(boolean tunnelProxy) {
     server.useHttps(sslClient.socketFactory, tunnelProxy);
   }
 
-  static final class DnsStart {
+  static class CallEvent {
```
Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: Hmm, not sure what I think of this. On the one hand it cuts down on the boilerplate and redundancy, but on the other hand we lose some type information on the params.

**Comment by dave-r12** [comment]
Re: connection release event -- I'm unsure on the intended use case there.

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java (hunk 1 above)
Comment (yschimke) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: There is no loss of type information as the fields are still toplevel.

The intent here was to implement as many of the asserts of matching pairs as part of the listener impl.  I'd like connectionReleased to fail fast if a matching connectionAcquired isn't already in place.  My goal was to incrementally implement the rules of the docs to make each new (edge case) unit test trivial to read and understand while having strong asserts.

**Comment by yschimke** [comment]
@dave-r12 Connection start/end pairs are great to show the cost of initial connections to the servers.  But in HTTP/1.1 especially there is a window where this call has exclusive ownership of a connection, so it's a very top level thing even when we redirect http->https and reuse existing connections.

**Comment by yschimke** [comment]
With connections shown.  n.b. I'll probably default to less detail in zipkin in my external PR, so this is just for demonstration.

image

image

**Comment by yschimke** [comment]
@dave-r12 @swankjesse I'm uncertain where to record responseBodyEnd and fetchEnd.  I suspect in Http1Codec and Http2Codec as they are internal classes, so API changes won't be disruptive like ResponseBody.

But it seems awkward at the moment, particularly with the number of special cases in Http1Codec e.g. chunked encoding and how to handle exceptions.  

The current PR will sometimes complete request body too early but is calling it reliably.

File: okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java [hunk 2]
```diff
@@ -82,26 +107,34 @@ public CallServerInterceptor(boolean forWebSocket) {
...
           .build();
     } else {
-      response = response.newBuilder()
-          .body(httpCodec.openResponseBody(response))
-          .build();
-    }
+      realChain.eventListener().responseBodyStart(realChain.call());
+      try {
+        response = response.newBuilder()
+            .body(httpCodec.openResponseBody(response))
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: I think we wanna move responseBodyStart() and responseBodyEnd() calls into the codec. Main reason is that if there’s exceptions reading the response body these will be the best places to signal the failure to responseBodyEnd.

File: okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    }
+    return eventTypes;
+  }
+
+  void clearAllEvents() {
+    eventSequence.clear();
+  }
+
+  private void logEvent(CallEvent e) {
+    CallEvent startEvent = e.closes();
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java]: This is neat

File: okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java (hunk 2 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: ALSO – I’d like the option of having request and response body events include the number of application-layer bytes in the message. Potentially useful for things like streaming JSON with gzip, where the response body size is pretty difficult to observe right now.

Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: Is it worth rethinking the EventListener and introduce parameter objects to allow additional fields in future without breaking compatibility?

Or marking the EventListener API as clearly experimental and likely to change?


```
  public void connectEnd(Call call, @Nullable ConnectEndDetails details, @Nullable Throwable throwable) {
  }

  public void responseBodyEnd(Call call, @Nullable ResponseBodyDetails details, @Nullable Throwable throwable) {
  }
```

Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: n.b. I moved the calls into the Codec, if it doesn't look right I can either

1) land as is and fix forward
2) revert to where I put it first time
3) fix after discussion

**Comment by yschimke** [comment]
Other tests

$ ./oksocial -z https://httpbin.org/stream-bytes/1073741824 > /dev/null

image

$ ./oksocial -z http://httpbin.org/redirect-to?url=http%3A%2F%2Fgoogle.com%2Frobots.txt

image

$ ./oksocial -z https://httpbin.org/status/500
HTTP/1.1 500 INTERNAL SERVER ERROR

File: okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java [hunk 4]
```diff
@@ -92,16 +117,16 @@ public CallServerInterceptor(boolean forWebSocket) {
       response = response.newBuilder()
           .body(httpCodec.openResponseBody(response))
           .build();
-    }
 
-    if ("close".equalsIgnoreCase(response.request().header("Connection"))
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: fixing this

**Comment by yschimke** [comment]
Suggest landing this as discussed and either marking as  experimental for next release or adopting a more forgiving pattern like parameter objects.

**Comment by yschimke** [comment]
Anything blocking this? Event code is in a weird state if you have to release without this anyway.

**Comment by dave-r12** [comment]
@yschimke the incomplete `EventListener` is my fault, although it was intentional. I was progressing through each event type so the CLs were smaller (I prefer small CLs to large CLs) and to bring focus on scrutinizing each event so everyone would be happy with the outcome. It wasn't clear to me how to make forward progress with this PR open. I didn't want to duplicate effort.

I can help get this completed.

**Comment by yschimke** [comment]
@dave-r12 feel free to take over this PR, or drop it and start from scratch.  I was trying to use the events and getting blocked on the missing ones.

**Comment by swankjesse** [comment]
Thanks @yschimke & @dave-r12. Sorry for the slow responses lately. Things have been pretty busy lately. I’ll look at this and try to get you both unblocked!

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java [hunk 5]
```diff
@@ -267,11 +267,11 @@ private void establishProtocol(ConnectionSpecSelector connectionSpecSelector, Ca
     eventListener.secureConnectStart(call);
     try {
       connectTls(connectionSpecSelector);
+      eventListener.secureConnectEnd(call, handshake, null);
     } catch (Exception e) {
       eventListener.secureConnectEnd(call, null, e);
       throw e;
     }
-    eventListener.secureConnectEnd(call, handshake, null);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: why move this? Only relevant if secureConnectEnd() throws

File: okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java [hunk 6]
```diff
@@ -301,6 +301,9 @@ private Socket deallocate(boolean noNewStreams, boolean released, boolean stream
       this.codec = null;
     }
     if (released) {
+      if (connection != null) {
+        eventListener.connectionReleased(call, connection);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java]: I’m a little anxious about calling user code while holding the single global connection pool lock. Probably safer to call this without the lock

File: okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java [hunk 7]
```diff
@@ -43,36 +43,61 @@ public CallServerInterceptor(boolean forWebSocket) {
...
-        responseBuilder = httpCodec.readResponseHeaders(true);
-      }
+      realChain.eventListener().requestBodyStart(realChain.call());
+      try {
+        // If there's a "Expect: 100-continue" header on the request, wait for a "HTTP/1.1 100
+        // Continue" response before transmitting the request body. If we don't get that, return
+        // what we did get (such as a 4xx response) without ever transmitting the request body.
+        if ("100-continue".equalsIgnoreCase(request.header("Expect"))) {
+          httpCodec.flushRequest();
+          // TODO event listener
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java]: ooh, tricky case

File: okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java [hunk 8]
```diff
@@ -129,10 +130,36 @@ public Http1Codec(OkHttpClient client, StreamAllocation streamAllocation, Buffer
   }
 
   @Override public ResponseBody openResponseBody(Response response) throws IOException {
-    Source source = getTransferStream(response);
+    streamAllocation.eventListener.responseBodyStart(streamAllocation.call);
+    Source source = new EventListenerSource(getTransferStream(response));
     return new RealResponseBody(response.headers(), Okio.buffer(source));
   }
 
+  class EventListenerSource extends ForwardingSource {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java]: I don’t think we want this as an extra object

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java]: I think it’s better without the extra abstraction

**Comment by swankjesse** [comment]
I merged! None of my feedback is urgent; just style stuff.

File: okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java (hunk 8 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java]: https://github.com/square/okhttp/pull/3502
